shaders/src/light_reflections.fs

```diff
@@ -0,0 +1,276 @@
+    P1 += vec2((distanceSquared(P0, P1) < 0.0001) ? 0.01 : 0.0);
+
+    Vector2 delta = P1 - P0;
+
+    // Permute so that the primary iteration is in x to reduce large branches later
+    bool permute = false;
+    if (abs(delta.x) < abs(delta.y)) {
+        // More-vertical line. Create a permutation that swaps x and y in the output
+        permute = true;
+        // Directly swizzle the inputs
+        delta = delta.yx;
+        P1 = P1.yx;
+        P0 = P0.yx;
+    }
+
+    // From now on, "x" is the primary iteration direction and "y" is the secondary one
+
+    float stepDirection = sign(delta.x);
+    float invdx = stepDirection / delta.x;
+    Vector2 dP = Vector2(stepDirection, invdx * delta.y);
+
+    // Track the derivatives of Q and k
+    Vector3 dQ = (Q1 - Q0) * invdx;
+    float   dk = (k1 - k0) * invdx;
+
+    // Scale derivatives by the desired pixel stride
+    dP *= stride; dQ *= stride; dk *= stride;
+
+    // Offset the starting values by the jitter fraction
+    P0 += dP * jitterFraction; Q0 += dQ * jitterFraction; k0 += dk * jitterFraction;
+
+    // Slide P from P0 to P1, (now-homogeneous) Q from Q0 to Q1, and k from k0 to k1
+    Point3 Q = Q0;
+    float  k = k0;
+
+    // We track the ray depth at +/- 1/2 pixel to treat pixels as clip-space solid voxels. Because
+    // the depth at -1/2 for a given pixel will be the same as at +1/2 for the previous iteration,
+    // we actually only have to compute one value per iteration.
+    float prevZMaxEstimate = vsOrigin.z;
+    float stepCount = 0.0;
+    float rayZMax = prevZMaxEstimate, rayZMin = prevZMaxEstimate;
+    float sceneZMax = rayZMax + 1e4;
+
+    // P1.x is never modified after this point, so pre-scale it by the step direction for a signed
+    // comparison
+    float end = P1.x * stepDirection;
+
+    // We only advance the z field of Q in the inner loop, since Q.xy is never used until after the
+    // loop terminates.
+
+    for (Point2 P = P0;
+        ((P.x * stepDirection) <= end) &&
+        (stepCount < maxSteps) &&
+        ((rayZMax < sceneZMax - vsZThickness) ||
+            (rayZMin > sceneZMax)) &&
+        (sceneZMax != 0.0);
+        P += dP, Q.z += dQ.z, k += dk, stepCount += 1.0) {
+
+        hitPixel = permute ? P.yx : P;
+
+        // The depth range that the ray covers within this loop iteration.  Assume that the ray is
+        // moving in increasing z and swap if backwards.  Because one end of the interval is shared
+        // between adjacent iterations, we track the previous value and then swap as needed to
+        // ensure correct ordering
+        rayZMin = prevZMaxEstimate;
+
+        // Compute the value at 1/2 pixel into the future
+        rayZMax = (dQ.z * 0.5 + Q.z) / (dk * 0.5 + k);
+        prevZMaxEstimate = rayZMax;
+        if (rayZMin > rayZMax) { swap(rayZMin, rayZMax); }
+
+        // View-space z of the background
+        sceneZMax = linearizeDepth(texelFetch(vsZBuffer, int2(hitPixel), 0).r);
```

Comment:
We need to explore if we could avoid the linearization here (future TODO) -- `sceneMax` is only used to check if we have a "hit", it's just a compare with the "z" of the ray, taking the thickness into account, those values could be "unlinearized" before the loop.  I'm not sure if it's feasible with `vsZThickness` though.